File: leetcode/回溯/93. 复原IP地址.py
from typing import List
class Solution:
    def restoreIpAddresses(self, s: str) -> List[str]:
        ans = []
        ip = []
        self.backtrack(s, ip, 0, ans)
        return ans

    def backtrack(self, remain, ip, part, ans):
        if part == 3:
            if remain != '':
                ip.append(remain)
                if self.legal_ip(ip):
                    ans.append('.'.join(ip))
                ip.pop()
        else:
            for i in range(0, 3):
                ip.append(remain[:i+1])
                self.backtrack(remain[i+1:], ip, part + 1, ans)
                ip.pop()

    def legal_ip(self, ip):
        for i in range(len(ip)):
            if i == 0:
                if not (0 < int(ip[i]) <= 255):
                    return False
            else:
                if not (0 <= int(ip[i]) <= 255):
                    return False
        return True


if __name__ == '__main__':
    s = "25525511135"
    solution = Solution()
    print(solution.restoreIpAddresses(s))

File: 每日一题/93. 复原IP地址.js
/**
 * @param {string} s
 * @return {string[]}
 * 
 * 主要是 合法 ip 字段的判断比较麻烦
 * 合法 ip: 除 0 外, 不得以 0 开头
 *
 */
var restoreIpAddresses = function(s) {
  if (s.length > 12) {
    return []
  }

  let ans = []
  backtracking(s, 0, [], 0, ans)
  return ans
};

function backtracking(s, start, prev, part, ans) {  // part 从 0 开始
  if (part == 4) {  
    if (start == s.length) {
      ans.push(prev.join('.'))
    }
    return
  }
  if (checkLen(s, start, part)) {
    for (let i = 1; i < 4; i++) {
      let end = start + i
      if (end > s.length) {
        break
      }
      if (validNum(s, start, end)) {
        prev.push(s.slice(start, end))
        backtracking(s, end, prev, part + 1, ans)
        prev.pop()
      }
    }
  }
}

function checkLen(s, start, part) {   // part 从 0 开始
  if (s.length - start > (4 - part) * 3) {
    return false
  }
  return true
}

function validNum(s, start, end) {
  let num = parseInt(s.slice(start, end))
  if(num == 0 && start + 1 != end){   // 排除 '00', '000'
    return false
  }
  if(num > 0 && s[start] == '0'){   // 排除 '01', '011', '001'
    return false
  }
  if (num >= 0 && num < 256) {
    return true
  }
  return false
}

console.log(restoreIpAddresses("010010"))